modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/logical/IgniteMultiJoinOptimizeBushyRule.java

```diff
@@ -0,0 +1,425 @@
+            best = bestSoFar;
+        }
+
+        RelNode result = relBuilder
+                .push(best.rel)
+                .filter(RexUtil.composeConjunction(rexBuilder, unusedConditions)
+                        .accept(new RexPermuteInputsShuttle(best.mapping, best.rel)))
+                .project(relBuilder.fields(best.mapping))
+                .build();
+
+        call.transformTo(result);
+    }
+
+    private static void aggregateEdges(Int2ObjectMap<List<Edge>> edges, int lhs, int rhs) {
+        int id = lhs | rhs;
+        if (!edges.containsKey(id)) {
+            Set<Edge> union = Collections.newSetFromMap(new IdentityHashMap<>());
+
+            union.addAll(edges.getOrDefault(lhs, List.of()));
+            union.addAll(edges.getOrDefault(rhs, List.of()));
+
+            edges.put(id, List.copyOf(union));
+        }
+    }
+
+    private static Vertex composeCartesianJoin(
+            Int2ObjectMap<Vertex> bestPlan,
+            Int2ObjectMap<List<Edge>> edges,
+            @Nullable Vertex bestSoFar,
+            RelMetadataQuery mq,
+            RelBuilder relBuilder,
+            RexBuilder rexBuilder
+    ) {
+        List<Vertex> options;
+
+        if (bestSoFar != null) {
+            options = new ArrayList<>();
+
+            for (Vertex option : bestPlan.values()) {
+                if ((option.id & bestSoFar.id) == 0) {
+                    options.add(option);
+                }
+            }
+        } else {
+            options = new ArrayList<>(bestPlan.values());
+        }
+
+        options.sort(VERTEX_COMPARATOR);
+
+        Iterator<Vertex> it = options.iterator();
+
+        if (bestSoFar == null) {
+            bestSoFar = it.next();
+        }
+
+        while (it.hasNext()) {
+            Vertex input = it.next();
+
+            if ((bestSoFar.id & input.id) != 0) {
+                continue;
+            }
+
+            List<Edge> edges0 = findEdges(bestSoFar.id, input.id, edges);
+
+            aggregateEdges(edges, bestSoFar.id, input.id);
+
+            bestSoFar = createJoin(bestSoFar, input, edges0, mq, relBuilder, rexBuilder);
+        }
+
+        return bestSoFar;
+    }
+
+    private static Vertex chooseBest(@Nullable Vertex currentBest, Vertex candidate) {
+        if (currentBest == null) {
+            return candidate;
+        }
+
+        if (VERTEX_COMPARATOR.compare(currentBest, candidate) > 0) {
+            return candidate;
+        }
+
+        return currentBest;
+    }
+
+    private static Int2ObjectMap<List<Edge>> collectEdges(LoptMultiJoin multiJoin, Set<RexNode> unusedConditions) {
+        Int2ObjectMap<List<Edge>> edges = new Int2ObjectOpenHashMap<>();
+
+        for (RexNode condition : multiJoin.getJoinFilters()) {
+            int[] inputRefs = multiJoin.getFactorsRefByJoinFilter(condition).toArray();
+
+            if (inputRefs.length < 2 || condition.isA(SqlKind.OR)) {
+                continue;
+            }
+
+            unusedConditions.remove(condition);
```

Comment:
We can fill `unusedConditions` inside the `if-block`  instead of removing them.
This will avoid one iteration over join filter, lookups into hashes, and Set can be replaced with a list.

Reply:
good point! Fixed